superbench/benchmarks/micro_benchmarks/directx_gemm_flops_performance/GPUCore.cpp

```diff
@@ -0,0 +1,507 @@
+    m_bindingTable->BindInputs(3, inputBindings.data());
+
+    DML_BUFFER_BINDING outputBufferBinding{m_outputBuffer.Get(), 0, sizeof(T) * n * m};
+    DML_BINDING_DESC outputBindingDesc{DML_BINDING_TYPE_BUFFER, &outputBufferBinding};
+
+    m_bindingTable->BindOutputs(1, &outputBindingDesc);
+
+    // Execute the compiled GEMM operator and record the GPU time.
+    this->gpuTimer.start(m_commandList.Get(), 0);
+    m_dmlCommandRecorder->RecordDispatch(m_commandList.Get(), m_dmlCompiledOperator.Get(), m_bindingTable.Get());
+    this->gpuTimer.stop(m_commandList.Get(), 0);
+    this->gpuTimer.resolveQueryToCPU(m_commandList.Get(), 0);
+    CloseExecuteResetWait();
+
+    double timeInMs = this->gpuTimer.getElapsedMsByTimestampPair(0);
+
+    if (_PRINT_RESULT) {
+        // The output buffer now contains the result of the identity operator,
+        // so read it back if you want the CPU to access it.
+
+        m_commandList->ResourceBarrier(
+            1, get_rvalue_ptr(CD3DX12_RESOURCE_BARRIER::Transition(
+                   m_outputBuffer.Get(), D3D12_RESOURCE_STATE_UNORDERED_ACCESS, D3D12_RESOURCE_STATE_COPY_SOURCE)));
+
+        m_commandList->CopyResource(m_readBackBuffer.Get(), m_outputBuffer.Get());
+
+        CloseExecuteResetWait();
+        D3D12_RANGE tensorBufferRange{0, static_cast<SIZE_T>(sizeof(T) * n * m)};
+        T *outputBufferData{};
+        ThrowIfFailed(m_readBackBuffer->Map(0, &tensorBufferRange, reinterpret_cast<void **>(&outputBufferData)));
+        std::wstring outputString = L"output tensor: ";
+        for (size_t tensorElementIndex{0}; tensorElementIndex < static_cast<SIZE_T>(m * n);
+             ++tensorElementIndex, ++outputBufferData) {
+            outputString += std::to_wstring(*outputBufferData) + L' ';
+        }
+
+        std::wcout << outputString << std::endl;
+        OutputDebugStringW(outputString.c_str());
+        D3D12_RANGE emptyRange{0, 0};
+        m_readBackBuffer->Unmap(0, &emptyRange);
+    }
+
+    return timeInMs;
+}
+
+/**
+ * @brief Close and execute command list, wait until command completed.
+ */
+void GPUCore::CloseExecuteResetWait() {
+    m_commandList->Close();
+    ID3D12CommandList *commandLists[] = {m_commandList.Get()};
+    m_commandQueue->ExecuteCommandLists(ARRAYSIZE(commandLists), commandLists);
+    FlushCommandQueue();
+    ThrowIfFailed(m_commandAllocator->Reset());
+    ThrowIfFailed(m_commandList->Reset(m_commandAllocator.Get(), nullptr));
+}
+
+/**
+ * @brief Wait until command completed.
+ */
+void GPUCore::FlushCommandQueue() {
+    // Signal and increment the fence value.
+    const UINT64 fenceL = m_currentFence;
+    m_commandQueue->Signal(m_fence.Get(), fenceL);
+    m_currentFence++;
+
+    // Wait until command queue is done.
+    if (m_fence->GetCompletedValue() < fenceL) {
+        m_fence->SetEventOnCompletion(fenceL, m_eventHandle);
+        WaitForSingleObject(m_eventHandle, INFINITE);
```

Comment:
>INFINITE [](http://example.com/codeflow?start=43&length=8)

INFINITE may cause the program hanging, so how about a time interval?

Reply:
add a 5 min limitation for default